tensorflow_datasets/video/moving_sequence.py

```diff
@@ -0,0 +1,419 @@
+      pad_left, = args
+      pad_right = total_padding - pad_left
+      padding = tf.stack([pad_left, pad_right], axis=-1)
+      z = tf.zeros((1, 2), dtype=pad_left.dtype)
+      padding = tf.concat([padding, z], axis=0)
+      return tf.pad(image, padding)
+
+    padded_images = tf.map_fn(
+      get_padded_image, [pad_lefts], dtype=tf.uint8, infer_shape=False,
+      back_prop=False)
+
+  return padded_images
+
+
+def create_moving_sequence(image, pad_lefts, total_padding):
+  """
+  Create a moving image sequence from the given image and left padding values.
+
+  Args:
+    image: [h, w, n_channels] uint8 array
+    pad_lefts: [sequence_length, 2] int32 array of
+      left padding values
+    total_padding: TensorShape or list/tuple, (out_h, out_w)
+
+  Returns:
+    [sequence_length, out_h, out_w, n_shannels] uint8 sequence.
+  """
+  total_padding = tf.TensorShape(total_padding)
+  pad_lefts = tf.convert_to_tensor(pad_lefts, dtype=tf.float32)
+  image = tf.convert_to_tensor(image, dtype=tf.uint8)
+  if image.shape.ndims != 3:
+    raise ValueError("`image` must be a rank 3 tensor")
+  if pad_lefts.shape.ndims != 2:
+    raise ValueError("`sequence_pad_lefts` must be a rank 2 tensor")
+  if len(total_padding) != 2:
+    raise ValueError(
+      "`total_padding` must have 2 entres, got %s"
+      % str(total_padding.as_list()))
+  seq = _create_moving_sequence(
+      image, pad_lefts, tf.convert_to_tensor(total_padding))
+  ph, pw = total_padding
+  h, w, n_channels = image.shape
+  sequence_length = pad_lefts.shape[0]
+  seq.set_shape((sequence_length, h + ph, w + pw, n_channels))
+  return seq
+
+
+def create_merged_moving_sequence(
+    images, sequence_pad_lefts, total_padding, background=tf.zeros,
+    merge_fn="max"):
+  """
+  Args:
+    images: [n_images, h, w, n_channels] uint8 array
+    sequence_pad_lefts: [n_images, sequence_length, 2] int32 array of
+      left padding values
+    total_padding: TensorShape (out_h, out_w)
+    background: background image, or callable that takes `shape` and `dtype`
+        args.
+    merge_fn: "max" for maximum, or callable mapping (seq0, seq1) -> seq, where
+        each of seq0, seq1 and seq2 aretensors of the same shape/dtype as
+        the output.
+
+  Returns:
+    [sequence_length, out_h, out_w, n_channels] overlayed padded sequence.
+  """
+  if isinstance(merge_fn, six.string_types):
+      merge_fn = _merge_fns[merge_fn]
+  images = tf.convert_to_tensor(images, dtype=tf.uint8)
+  sequence_pad_lefts = tf.convert_to_tensor(sequence_pad_lefts, dtype=tf.int32)
+  if images.shape.ndims != 4:
+    raise ValueError("`images` must be a rank 4 tensor")
+  if sequence_pad_lefts.shape.ndims != 3:
+    raise ValueError("`sequence_pad_lefts` must be a rank 4 tensor")
+  if len(total_padding) != 2:
+    raise ValueError(
+      "`total_padding` must be len 2, got %s"
+      % str(total_padding.as_list()))
+
+  image_res = [i + t for i, t in zip(images.shape[1:3], total_padding)]
```

Comment:
Image shape required to be fully known statically? If so, check and raise an error up front. Or make it so that it can be dynamic.